src/server/pps/server/s3g_sidecar.go

```diff
@@ -0,0 +1,359 @@
+
+	// End runs when a job ends
+	End func(jobID string) error
+}
+
+func (s *sidecarS3G) HandleJobs(h JobHandler) error {
+	retryCtx, retryCancel := context.WithCancel(context.Background())
+	defer retryCancel()
+	watcher, err := s.apiServer.jobs.ReadOnly(retryCtx).WatchByIndex(ppsdb.JobsPipelineIndex, s.pipelineInfo.Pipeline)
+	if err != nil {
+		return fmt.Errorf("error creating watch: %v", err)
+	}
+	defer watcher.Close()
+	for e := range watcher.Watch() {
+		jobID := string(e.Key)
+		if e.Type == watch.EventError {
+			return fmt.Errorf("sidecar s3 gateway watch error: %v", e.Err)
+		} else if e.Type == watch.EventDelete {
+			// Job was deleted, e.g. because input commit was deleted. Note that the
+			// service may never have been created (see IsErrNotFound under InspectJob
+			// below), so no error is returned if jobID is not in s.servers
+			if h.End != nil {
+				if err := h.End(jobID); err != nil {
+					return err
+				}
+			}
+			continue
+		}
+		// 'e' is a Put event (new or updated job)
+		// create new ctx for this job, and don't use retryCtx as the
+		// parent. Just because another job's etcd write failed doesn't
+		// mean this job shouldn't run
+		jobCtx, jobCancel := context.WithCancel(s.pachClient.Ctx())
+		defer jobCancel() // cancel the job ctx
+		pachClient := s.pachClient.WithCtx(jobCtx)
+		// Inspect the job and make sure it's relevant, as this worker may be old
+		logrus.Errorf("sidecar s3 gateway: inspecting job %q to begin serving inputs over s3 gateway", jobID)
+		jobInfo, err := pachClient.InspectJob(jobID, false)
+		if err != nil {
+			if col.IsErrNotFound(err) {
+				// TODO(msteffen): I'm not sure what this means--maybe that the service
+				// was created and immediately deleted, and there's a pending deletion
+				// event? In any case, without input commit IDs there's nothing to do
+				continue
+			}
+			return fmt.Errorf("error from InspectJob(%v): %+v", jobID, err)
+		}
+		if jobInfo.PipelineVersion < s.pipelineInfo.Version {
+			logrus.Infof("skipping job %v as it uses old pipeline version %d", jobID, jobInfo.PipelineVersion)
+			continue
+		}
+		if jobInfo.PipelineVersion > s.pipelineInfo.Version {
+			return fmt.Errorf("job %s's version (%d) greater than pipeline's "+
+				"version (%d), this should automatically resolve when the worker "+
+				"is updated", jobID, jobInfo.PipelineVersion, s.pipelineInfo.Version)
+		}
+		if ppsutil.IsTerminal(jobInfo.State) {
+			if h.End != nil {
+				if err := h.End(jobID); err != nil {
+					return err
+				}
+			}
+			continue
+		}
+
+		if h.Begin != nil {
+			if err := h.Begin(jobCtx, jobInfo); err != nil {
+				return err
+			}
+		}
+	}
+	return fmt.Errorf("job watcher loop was broken out of; this should not happen")
+}
+
+func (s *sidecarS3G) Client(authToken string) (*client.APIClient, error) {
+	newClient := s.apiServer.env.GetPachClient(s.pachClient.Ctx()) // clones s.pachClient
+	newClient.SetAuthToken(authToken)
```

Comment:
This is presumably to match the client factory interface? A couple of things here:

1) I should remove the `authToken` parameter, and just call `SetAuthToken` on the client that's returned. It'll simplify the several implementations of client factory.
2) Got a recommended name instead of client factory? It's embarassingly java-esque :D

Reply:
Added a comment clarifying that yes, that is the goal :-).

I actually wonder if it might make sense for the s3 gateway to take `serviceenv.ServiceEnv` as a parameter, which implements `GetPachClient(ctx)`. Every pach client it returns is a clone, so there's no worry that calling `SetAuthToken` will somehow interfere with other code elsewhere.

Reply:
Is there a way to trivially use `ServiceEnv` in tests? I'd need that for s3 integration testing.

Reply:
Not yet, though I've discussed this with @Tryneus a bit. The goal is to create an interface that both `serviceenv.ServiceEnv` and `testing.RealEnv` satisfy, that can be passed in to all APIServers that need a client. We can punt on passing a `ServiceEnv` to s3 gateway until then.

It's a bit unfortunate that we have two data structures whose essential jobs are to pass a pachyderm client, but once we have a standard interface for both of them, I think there are many parts of our code that will be able to accept mock environments.